File: src/dlinkedList.cpp
#include "dlinkedList.h"
#include <cstring>
#include <fstream>
#include <iostream>

using namespace std;

// Инициализация списка.
void DlinkedList::init() {
    head = nullptr;
    tail = nullptr;
}

// Добавление нового элемента в начало.
void DlinkedList::addToHead(const string& value) {
    DlistNode* newNode = new DlistNode{value, head, nullptr};
    if (head != nullptr) {
        head->prev = newNode;
    }
    head = newNode;
    if (tail == nullptr) {
        tail = head;
    }
}

// Добавление нового элемента в конец.
void DlinkedList::addToTail(const string& value) {
    DlistNode* newNode = new DlistNode{value, nullptr, tail};
    if (tail != nullptr) {
        tail->next = newNode;
    }
    tail = newNode;
    if (head == nullptr) {
        head = tail;
    }
}

// Удаление элемента с головы.
void DlinkedList::removeFromHead() {
    if (head == nullptr) {
        return;
    }
    DlistNode* temp = head;
    head = head->next;
    if (head != nullptr) {
        head->prev = nullptr;
    } else {
        tail = nullptr;
    }
    delete temp;
}

// Удаление элемента с хвоста.
void DlinkedList::removeFromTail() {
    if (tail == nullptr) {
        return;
    }
    DlistNode* temp = tail;
    tail = tail->prev;
    if (tail != nullptr) {
        tail->next = nullptr;
    } else {
        head = nullptr;
    }
    delete temp;
}

// Удаление узла по значению.
void DlinkedList::removeByValue(const string& value) {
    DlistNode* temp = head;
    while (temp != nullptr) {
        if (temp->data == value) {
            if (temp->prev != nullptr) {
                temp->prev->next = temp->next;
            } else {
                head = temp->next;
            }
            if (temp->next != nullptr) {
                temp->next->prev = temp->prev;
            } else {
                tail = temp->prev;
            }
            delete temp;
            return;
        }
        temp = temp->next;
    }
}

// Поиск элемента по значению.
bool DlinkedList::search(const string& value) {
    DlistNode* temp = head;
    while (temp != nullptr) {
        if (temp->data == value) {
            return true;
        }
        temp = temp->next;
    }
    return false;
}

// Вывод всех элементов списка.
void DlinkedList::print() {
    DlistNode* temp = head;
    while (temp != nullptr) {
        cout << temp->data << " ";
        temp = temp->next;
    }
    cout << endl;
}

// Очистка списка.
void DlinkedList::destroy() {
    while (head != nullptr) {
        removeFromHead();
    }
}

// Загрузка элементов из файла.
void DlinkedList::loadFromFile(const string& fileName) {
    ifstream file(fileName);
    string value;
    while (file >> value) {
        addToTail(value);
    }
    file.close();
}

// Сохранение элементов списка в файл.
void DlinkedList::saveToFile(const string& fileName) {
    ofstream file(fileName);
    DlistNode* temp = head;
    while (temp != nullptr) {
        file << temp->data << endl;
        temp = temp->next;
    }
    file.close();
}

// Функция запуска работы со списком, обработка команд из аргументов командной строки.
void runLLinkedList(int argc, char* argv[]) {
    DlinkedList list;
    list.init();

    string fileName;
    string query;

    for (int i = 1; i < argc; i++) {
        if (strcmp(argv[i], "--file") == 0 && i + 1 < argc) {
            fileName = argv[i + 1];
            i++;
        } else if (strcmp(argv[i], "--query") == 0 && i + 1 < argc) {
            query = argv[i + 1];
            i++;
        }
    }

    list.loadFromFile(fileName);

    string command;
    size_t pos = query.find(' ');
    if (pos != string::npos) {
        command = query.substr(0, pos);
        query = query.substr(pos + 1);
    } else {
        command = query;
    }

    if (command == "DPUSH") {
        list.addToHead(query);  // Добавление элемента в начало списка
        list.saveToFile(fileName);
    } else if (command == "DAPPEND") {
        list.addToTail(query);  // Добавление элемента в конец списка
        list.saveToFile(fileName);
    } else if (command == "DREMOVEHEAD") {
        list.removeFromHead();  // Удаление элемента с головы списка
        list.saveToFile(fileName);
    } else if (command == "DREMOVETAIL") {
        list.removeFromTail();  // Удаление элемента с конца списка
        list.saveToFile(fileName);
    } else if (command == "DREMOVE") {
        list.removeByValue(query);  // Удаление элемента по значению
        list.saveToFile(fileName);
    } else if (command == "DSEARCH") {
        cout << (list.search(query) ? "true" : "false") << endl;    // Поиск элемента по значению
    } else if (command == "DPRINT") {
        list.print();
    }

    list.destroy();
}

File: Go/src/dlinkedList.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

// DlistNode — узел двусвязного списка
type DlistNode struct {
	data string
	next *DlistNode
	prev *DlistNode
}

// DlinkedList — двусвязный список
type DlinkedList struct {
	head *DlistNode
	tail *DlistNode
}

func (l *DlinkedList) Init() {
	l.head = nil
	l.tail = nil
}

func (l *DlinkedList) AddToHead(value string) {
	newNode := &DlistNode{data: value, next: l.head, prev: nil}
	if l.head != nil {
		l.head.prev = newNode
	}
	l.head = newNode
	if l.tail == nil {
		l.tail = l.head
	}
}

func (l *DlinkedList) AddToTail(value string) {
	newNode := &DlistNode{data: value, next: nil, prev: l.tail}
	if l.tail != nil {
		l.tail.next = newNode
	}
	l.tail = newNode
	if l.head == nil {
		l.head = l.tail
	}
}

func (l *DlinkedList) RemoveFromHead() {
	if l.head == nil {
		return
	}
	temp := l.head
	l.head = l.head.next
	if l.head != nil {
		l.head.prev = nil
	} else {
		l.tail = nil
	}
	_ = temp
}

func (l *DlinkedList) RemoveFromTail() {
	if l.tail == nil {
		return
	}
	temp := l.tail
	l.tail = l.tail.prev
	if l.tail != nil {
		l.tail.next = nil
	} else {
		l.head = nil
	}
	_ = temp
}

func (l *DlinkedList) RemoveByValue(value string) {
	for cur := l.head; cur != nil; cur = cur.next {
		if cur.data == value {
			if cur.prev != nil {
				cur.prev.next = cur.next
			} else {
				l.head = cur.next
			}
			if cur.next != nil {
				cur.next.prev = cur.prev
			} else {
				l.tail = cur.prev
			}
			return
		}
	}
}

func (l *DlinkedList) Search(value string) bool {
	for cur := l.head; cur != nil; cur = cur.next {
		if cur.data == value {
			return true
		}
	}
	return false
}

func (l *DlinkedList) Print() {
	first := true
	for cur := l.head; cur != nil; cur = cur.next {
		if !first {
			fmt.Print(" ")
		}
		first = false
		fmt.Print(cur.data)
	}
	fmt.Println()
}

func (l *DlinkedList) Destroy() {
	l.head = nil
	l.tail = nil
}

func (l *DlinkedList) LoadFromFile(fileName string) error {
	f, err := os.Open(fileName)
	if err != nil {
		return err
	}
	defer f.Close()
	scanner := bufio.NewScanner(f)
	for scanner.Scan() {
		l.AddToTail(scanner.Text())
	}
	return scanner.Err()
}

func (l *DlinkedList) SaveToFile(fileName string) error {
	f, err := os.Create(fileName)
	if err != nil {
		return err
	}
	defer f.Close()
	w := bufio.NewWriter(f)
	for cur := l.head; cur != nil; cur = cur.next {
		_, _ = w.WriteString(cur.data + "\n")
	}
	return w.Flush()
}

func runLLinkedList(fileName string, fullQuery string) {
	list := DlinkedList{}
	list.Init()
	_ = list.LoadFromFile(fileName) // если файла нет — продолжим с пустым списком

	parts := strings.SplitN(fullQuery, " ", 2)
	command := parts[0]
	arg := ""
	if len(parts) > 1 {
		arg = parts[1]
	}

	switch command {
	case "DPUSH":
		list.AddToHead(arg)
		_ = list.SaveToFile(fileName)
	case "DAPPEND":
		list.AddToTail(arg)
		_ = list.SaveToFile(fileName)
	case "DREMOVEHEAD":
		list.RemoveFromHead()
		_ = list.SaveToFile(fileName)
	case "DREMOVETAIL":
		list.RemoveFromTail()
		_ = list.SaveToFile(fileName)
	case "DREMOVE":
		list.RemoveByValue(arg)
		_ = list.SaveToFile(fileName)
	case "DSEARCH":
		if list.Search(arg) {
			fmt.Println("true")
		} else {
			fmt.Println("false")
		}
	case "DPRINT":
		list.Print()
	default:
		fmt.Fprintln(os.Stderr, "Unknown dlinkedlist command:", command)
	}

	list.Destroy()
}
